### ISSUE TITLE
[FEATURE]: Ensembling 

### ISSUE THREAD

**Issue by KnathanM** [chemprop/chemprop#581]
**Use-cases/examples of this new feature**
Many models achieve lower error when using the average of several models trained from different initializations. We want to add the ability to request the model to be an ensemble of models. 

**Desired solution/workflow**
In v1, users could just use the `--ensemble_size` command. It was also related to cross validation. In at least one paper I read they used 5 data splits and 5 weight initializations for parameters, leading to an ensemble of 25 models. It would be possible to also have 25 weight initializations on 1 data split to get an ensemble of the same size.

**Comment by davidegraff** [comment]
We should think about two things:
1. training: should we always train under the assumption of ensembling? That is, training a single model can be framed as an ensemble size of `k=1`. Alternatively, we can have two different modes: mono- and multi-model training. The benefit of training only a single model is that the file structure is flatter, but if we always assume ensembling, then the file structure is consistent. We will also have to consider the implications of cross-validation on the resulting file structure. My vote is to assume ensembling with a default size of 1 to produce the following general file structure _without_ CV for
`--ensemble-size=j`:
```
output_no_cv/
├── model_0/
│   ├── model.ckpt
│   ├── preds.csv
│   └── scaler.pkl
[...]
└── model_j/
    ├── model.ckpt
    ├── preds.csv
    └── scaler.pkl
```
and _with_ `k`-fold cross-validation:
```
output_with_cv/
├── model_0/
│   ├── fold_0/
│   │   ├── model.ckpt
│   │   ├── preds.csv
│   │   └── scaler.pkl
│ [...]
│   └── fold_k/
[...]
└── model_j/
    ├── fold_0
  [...]
    └── fold_k
```
2. prediction: provided some abstract directory that is some combination of ensembling and cross-validation, how should we load all of these models and then make predictions?

**Comment by kevingreenman** [comment]
I agree about assuming ensembling with a default size of 1 to give consistent file structure. @davidegraff is there a reason you suggest not also assuming 1-fold cross-validation? Chemprop v1 assumed both and had the `/fold_0/model_0/` structure. We can certainty deviate from this if we want to for v2, but it might be better to stick with the original convention unless we have a good reason to change it.

**Comment by davidegraff** [comment]
Well I don't think the concept of "1-fold cross-validation" is correct for standard model training. That is, in $k$-fold CV, you partition the data into $k$ separate folds and, for each fold, train a model on the reamining $\frac{k-1}{k}$ of the data. For $k=1$, this means 1 testing fold and training a model on no data ($\frac{1-1}{1}$). I understand the motivation here for maintaining parallel directory structures between v1 and v2 as well as multiple modes of v2 training, but the application of the concept here isn't correct IMO.

**Comment by kevingreenman** [comment]
Ah I see what you mean. Regardless of the structure we decide on, we should also think about how ensembles / CV will impact loading / converting models trained in v1 to predict using v2.

**Comment by hwpang** [comment]
> Well I don't think the concept of "1-fold cross-validation" is correct for standard model training. That is, in k-fold CV, you partition the data into k separate folds and, for each fold, train a model on the remaining k−1k of the data. For k=1, this means 1 testing fold and training a model on no data (1−11). I understand the motivation here for maintaining parallel directory structures between v1 and v2 as well as multiple modes of v2 training, but the application of the concept here isn't correct IMO.

I'm working on https://github.com/chemprop/chemprop/pull/608 to try to implement cross validation and ensembling for v2. I think there are two kinds of cross validation. The first one is what @davidegraff just described, and the second kind, which is what chemprop v1 does, is that it samples randomly with different random state for each fold (see [here](https://github.com/chemprop/chemprop/blob/master/chemprop/train/cross_validate.py#L106)), so 1-fold cross validation makes sense with what chemprop v1 does. This is also what the split_data function does in v2.

Based on this, I think it makes sense to keep the `/fold_0/model_0/` from v1. What do you think?

**Comment by davidegraff** [comment]
> I'm working on #608 to try to implement cross validation and ensembling for v2. I think there are two kinds of cross validation. The first one is what @davidegraff just described, and the second kind, which is what chemprop v1 does, is that it samples randomly with different random state for each fold (see [here](https://github.com/chemprop/chemprop/blob/master/chemprop/train/cross_validate.py#L106)), so 1-fold cross validation makes sense with what chemprop v1 does. This is also what the split_data function does in v2.
> 
> Based on this, I think it makes sense to keep the `/fold_0/model_0/` from v1. What do you think?

I disagree. The canonical definition of cross-validation, and indeed the one you will find upon searching "cross validation," is broadly defined as (1) partitioning the data into subsets, (2) for each subset of your data, train your model on the union of all remaining subsets and test your model on the given subset, (3) report the average test performance of each model[^1][^2][^3]. We can't redefine "cross validation" just so it fits with our desired object model. In fact, I can't even tell what v1 was doing because there are two splitting modes that ostensibly both correspond to cross-validation, `"cv"` and `"crossval"`, where the former [doesn't allow for values of `k=1` ](https://github.com/chemprop/chemprop/blob/54bd7b376960ed34f5b678dc00f5fc3f13e87f50/chemprop/data/utils.py#L710)

[^1]: https://en.wikipedia.org/wiki/Cross-validation_(statistics)#k-fold_cross-validation
[^2]: https://scikit-learn.org/stable/modules/cross_validation.html
[^3]: https://machinelearningmastery.com/k-fold-cross-validation/

**Comment by hwpang** [comment]
That's good points. Thanks for the detailed explanation. The current "cv" option we provide in the [`split_data`](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/data/splitting.py#L97) function in v2 also takes this resampling approach I described above. We might want to discuss which we want to support in v2 and change the `split_data` function accordingly.

**Comment by hwpang** [comment]
Regarding the output directory structure of cross validation, what about this:

- For the case we do use k-fold cross-validation (number of data points > k >= 2), we have output directory structure of `/fold_i/molde_0/` for all i in {0,...,k-1}
- For the case we don't use cross-validation, we have the output directory structure of `/molde_0/`?

**Comment by davidegraff** [comment]
resampling seems to be just model ensembling then, no? In that, if I wanted an ensemble of `j` models to lower overall model variance given some dataset, I would likely resample the dataset to create `j` train/val/test splits and get `j` different models.

**Comment by hwpang** [comment]
Yeah, I think you're right. After thinking about this, here are some thoughts on how we might improve the cross-validation in `data_split` while also considering model ensembling.

Option A: (1) let `CV` option in `data_split` returns the $k$-fold data. For $k$-fold cross-validation, it returns a list of $k$ lists. For each ensemble model, 1 out of $k$ is held out as the test set, and uses `RANDOM_NO_VAL` on the remaining $\frac{k-1}{k}$ to obtain the training set and the validation set.

Option B: the `CV` option in `data_split` takes both `num_folds` and `ensemble_size` as input, and split the data based on A(1) and A(2), and return a list of lists of lists, where the second layer has $k$ lists, and the third layer has 3 lists, corresponding to training, validation, and testing.

What do you think?

**Comment by hwpang** [comment]
This is closed by https://github.com/chemprop/chemprop/pull/631.
